WordPress/wordpress-playground
# [Cors Proxy] Support Transfer-Encoding: Chunked

Adjusts the CORS Proxy to relay chunked responses using chunks.

 ## Implementation

Some APIs use fixed content-length and respond with a predictable response body. Other APIs respond with chunks.

Before this PR, whenever the CORS proxy received a chunked response, it would only relay the `Transfer-Encoding: Chunked` header without actually transmitting the response body as chunks.

Chunks are an essential HTTP 1.1 feature that cannot be disabled, We also can't just filter out that header in the proxy response as the client expects either a Content-Length or body chunks, and we may not know the Content-length upfront.

With this PR, the CORS proxy outputs its response using the chunked encoding whenever the remote endpoint streams its response using the chunked encoding.

 ## Testing

* Run `nx start playground-php-cors-proxy`
* Fixed-length response: Go to http://127.0.0.1:5263/cors-proxy.php?https://raw.githubusercontent.com/WordPress/wordpress-playground/5e5ba3e0f5b984ceadd5cbe6e661828c14621d25/README.md
* Confirm it loaded the file
* Chunked encoding: Go to http://127.0.0.1:5263/cors-proxy.php?https://api.wordpress.org/plugins/update-check/1.1/
* Confirm the response says the word `error`
* Error handling: Go to http://127.0.0.1:5263/cors-proxy.php?https://nosuchsite.playground.wordpress.net
* Confirm the response says Bad Request Hostname could not be resolved
